common/type_system/TypeSystem.cpp

```diff
@@ -70,6 +70,46 @@ Type* TypeSystem::add_type(const std::string& name, std::unique_ptr<Type> type)
   return m_types[name].get();
 }
 
+/*!
+ * Add a new 'enum type'. This maps enum names to the their type's name, allowing the enum name to
+ * be used as if it were a type name.
+ */
+Type* TypeSystem::add_enum_type(const std::string& name, const std::string& type) {
+  auto t = lookup_type(type);
+  assert(t);
+  auto kv = m_enum_types.find(name);
+  if (kv != m_enum_types.end()) {
+    // exists already
+
+    if (kv->second != type) {
+      // exists, and we are trying to change it!
+      fmt::print("[TypeSystem] Enum {} was originally\n{}\nand is redefined as\n{}\n", name,
+                 kv->second, type);
+
+      throw std::runtime_error("Enum type was redefined.");
+    }
+  } else {
+    // Should forward declaring enums be forbidden? would it even necessarily break anything? I'll
+    // keep it allowed for now, though it is a weird thing to do...
```

Comment:
Let's make this an error. I don't think there's a reason to forward declare enums, and we want to prevent the case where somebody forward declares a type as a structure, and then uses that same name for an enum.